palette/src/yuv/quant.rs

```diff
@@ -0,0 +1,55 @@
+use super::QuantizationFn;
+use {clamp, cast, Component, Float};
+
+/// Shared 8-bit quantization functions.
+pub struct QuantU8;
+
+impl QuantU8 {
+    fn quantize_yuv<F: Component + Float>([y, u, v]: [F; 3]) -> [u8; 3] {
+        let y = y*cast(219.) + cast(16.);
+        let u = u*cast(224.) + cast(128.);
+        let v = v*cast(224.) + cast(128.);
+        [int_u8(y), int_u8(u), int_u8(v)]
+    }
+
+    fn quantize_rgb<F: Component + Float>([r, g, b]: [F; 3]) -> [u8; 3] {
+        let r = r*cast(219.) + cast(16.);
+        let g = g*cast(219.) + cast(16.);
+        let b = b*cast(219.) + cast(16.);
+        [int_u8(r), int_u8(g), int_u8(b)]
+    }
+}
+
+impl QuantizationFn for QuantU8 {
+    type Output = u8;
+
+    fn quantize_yuv<F: Component + Float>(yuv: [F; 3]) -> [u8; 3] {
+        Self::quantize_yuv(yuv)
+    }
+
+    fn quantize_rgb<F: Component + Float>(rgb: [F; 3]) -> [u8; 3] {
+        Self::quantize_rgb(rgb)
+    }
+}
+
+/// Round to 8-bit integer in valid signal range.
+fn int_u8<F: Component + Float>(value: F) -> u8 {
+    // Note: signal level below 1 and the level 255 and above are reserved.
+    clamp(value.round(), cast(1.), cast(254.)).convert()
+}
+
+// TODO: 10bit quantization. Already here as a reference for encoding to 10 bit quantized values.
+#[allow(unused)]
+fn to_u10<F: Component + Float>(value: F) -> u16 {
+    // Representation of 254.75 with 2 fractional bits.
+    const UPPER_MAX: u16 = 4*254 + 3; // = 1019
+    // Representation of 1.0 with 2 fraction bits.
+    const LOWER_MIN: u16 = 4;
+
+    // Add two fractional bits.
+    let value = value*cast(4.);
+
+    // Note: signal level below 1 and the level 255 and above are reserved.
+    clamp(value.round(), cast(LOWER_MIN), cast(UPPER_MAX)).convert()
```

Comment:
This `convert` is also questionable...

Reply:
And the comment mentions 255. :slightly_smiling_face:

Reply:
Is `convert` the scaling one? I suppose both this and the one in `u8` should instead be `NumCast::from`

Reply:
it's meant to scale between a 0.0 - 1.0 scale and any unsigned integer (where it will span the full value range), so it's going to nomalize it to 0.0 - 0.1 and then scale it back to 0 - 255. You can just use `NumCast::from` or `cast` (which is basically the same idea), but you may also want to round it. I don't know if it's supposed to be rounded or floored in this case.

Reply:
Or it may not even normalize it, since it's already a float. It may actually just output garbage here. :astonished: It's going to run [this one](https://github.com/Ogeon/palette/blob/master/palette/src/lib.rs#L549) or the one for f64, if I'm reading it correctly, so it multiplies with 255.

Reply:
Ah, then `NumCast::from` is the one to take. The rounding is supposed to already have been applied to the `F: Float` itself, and it should be an actual rounding with `.5` upwards (that's actually not completely true for IEEE floats afaik but good enough in this range since the half will never be the last bit of the number).

Reply:
(Never mind about the rounding, I somehow missed that you are already doing that)

Reply:
But yes, you want to just cast it in this case. The naming could be better ("convert" is very unspecific), but I think I had a hard time coming up with something that didn't make it worse. :laughing: